File: code/frontend/src/lib/lexEngine/generate.ts
/**
 * Generation for the Lexicon Lab.
 *
 * Greedy at `temperature = 0`, otherwise a sample from `softmax(logits[-1] / T)` after
 * setting the logits of `GENERATION_BANNED_IDS = (<unk>, <bos>, <pad>)` to -inf. `<eos>`
 * IS sampleable — it is how the model ends a line.
 *
 * **In-budget output is guaranteed by construction**, because the vocabulary IS the
 * budget: there are no rows for words outside it, so there is nothing to filter. No
 * trie, no post-filter, and none of the source project's `" hameat"` defect (its trie
 * re-opened the root after a completed word). FR-605 / SC-602.
 *
 * "By construction" holds only while the arithmetic does, so it is CHECKED, not assumed.
 * A model whose weights have left float32 range (a `×2` preset applied often enough, or
 * one cell edit of `1e40`, which is finite in JS and `Infinity` once stored) produces NaN
 * logits, and `NaN > -Infinity` is false — so the greedy argmax used to fall through to
 * its initialiser `next = 0`, which is `<unk>`, bypassing the -inf ban entirely, and the
 * sampler fell through to `V - 1`. Both were then hidden by stripping specials from
 * `words`. Non-finite weights and non-finite logits now raise `ComputeError`, exactly as
 * `lex/generate.py` does; nothing is clamped and nothing is sanitised.
 */

import { computeError, invalidParam } from "../geoEngine/errors";
import { LexModel, nonFiniteWeightNames, sfc32 } from "./model";
import { BOS_ID, EOS_ID, GENERATION_BANNED_IDS, LexVocab, SPECIAL_TOKENS, tokenize } from "./vocab";

export const DEFAULT_TEMPERATURE = 0.9;
export const DEFAULT_MAX_NEW_TOKENS = 40;
export const MAX_NEW_TOKENS = 200;

export interface GenerateOptions {
  /** Free text; tokenized with the model's own vocabulary. Out-of-budget words become
   *  `<unk>` and are reported in `promptOov` rather than silently dropped. */
  prompt?: string;
  maxNewTokens?: number;
  /** 0 = greedy (argmax). Above 0, sample from the tempered distribution. */
  temperature?: number;
  seed?: number;
  /**
   * What `<eos>` means. Defaults to `false`, matching the backend's
   * `generate(..., stop_at_eos=False)`: the model runs for the full `maxNewTokens` and
   * `<eos>` renders as a LINE BREAK, which is how a nursery-rhyme model produces verse
   * rather than one long line. `true` stops at the first `<eos>` for a single line.
   *
   * This existed only as the `true` behaviour, which made multi-line output reachable on
   * the full stack and not in the browser — the two runtimes have to mean the same thing
   * by the same name.
   */
  stopAtEos?: boolean;
}

export interface GenerateResult {
  /** The newly generated ids only, `<eos>` included when the model chose to stop. */
  ids: number[];
  /**
   * The generated words as a FLAT list, with the line breaks dropped.
   *
   * `<eos>` is the only token this can remove: `<unk>`, `<bos>` and `<pad>` are masked to
   * -inf before every draw and their appearance is treated as a bug (it raises), so this
   * is a rendering choice — the same ids as `text` without its newlines — and not a
   * budget filter. Every entry is in budget because the vocabulary is the budget.
   */
  words: string[];
  text: string;
  promptIds: number[];
  /** Prompt words the budget cannot express (they entered the model as `<unk>`). */
  promptOov: string[];
  stoppedOnEos: boolean;
}

const SPECIALS = new Set<string>(SPECIAL_TOKENS);
const BANNED = new Set<number>(GENERATION_BANNED_IDS);

/**
 * The masked logits must be usable before anything reads them: at least one row still
 * finite (the backend's `torch.isfinite(logits).any()`), and no NaN or +Infinity anywhere
 * (which the backend catches as non-finite probabilities, and which no comparison-based
 * argmax can survive). Both raise; neither is repaired.
 */
function assertUsableLogits(logits: Float64Array, V: number, step: number): void {
  let anyFinite = false;
  for (let i = 0; i < V; i++) {
    const x = logits[i];
    if (Number.isNaN(x) || x === Infinity) {
      throw computeError(
        `the model produced non-finite logits at step ${step + 1}, so no continuation is ` +
          "defined. This means the weights have left float32 range — most likely a weight " +
          "edit (values above ~3.4e38 become Infinity when stored) or a diverged run. " +
          "Refusing to generate rather than emit an arbitrary token.",
      );
    }
    if (Number.isFinite(x)) anyFinite = true;
  }
  if (!anyFinite) {
    throw computeError(
      `all logits are -inf at step ${step + 1}; generation cannot continue`,
    );
  }
}

export function generate(model: LexModel, vocab: LexVocab, opts: GenerateOptions = {}): GenerateResult {
  const maxNew = Math.trunc(opts.maxNewTokens ?? DEFAULT_MAX_NEW_TOKENS);
  const temperature = opts.temperature ?? DEFAULT_TEMPERATURE;
  const stopAtEos = opts.stopAtEos ?? false; // backend default; see GenerateOptions
  if (!(maxNew >= 1 && maxNew <= MAX_NEW_TOKENS)) {
    throw invalidParam(`max_new_tokens must be in 1..${MAX_NEW_TOKENS}, got ${opts.maxNewTokens}`);
  }
  if (!(temperature >= 0 && Number.isFinite(temperature))) {
    throw invalidParam(`temperature must be a finite number >= 0, got ${opts.temperature}`);
  }
  if (vocab.rows !== model.cfg.vocabRows) {
    throw invalidParam(
      `this vocabulary has ${vocab.rows} rows but the model has ${model.cfg.vocabRows} — ` +
        "a model must be generated with the vocabulary it was trained on",
    );
  }

  // Refuse a broken model up front, where the cause is still nameable. `lex/generate.py`
  // reaches the same refusal one step later, from its logits.
  const badWeights = nonFiniteWeightNames(model.cfg, model.weights);
  if (badWeights.length > 0) {
    throw computeError(
      `this model's weights are not finite: ${badWeights.join(", ")} contain NaN or ±Infinity, ` +
        "most likely from a weight edit that pushed them out of float32 range (values above " +
        "~3.4e38 become Infinity when stored) or from a diverged training run. Generation " +
        "would be meaningless, so it is refused rather than approximated.",
    );
  }

  const promptTokens = tokenize(opts.prompt ?? "");
  const promptOov = promptTokens.filter((t) => !vocab.has(t));
  const promptIds = vocab.encode(promptTokens);

  const rand = sfc32(Math.trunc(opts.seed ?? 0));
  const V = model.cfg.vocabRows;
  const ctx = model.cfg.ctx;
  const context: number[] = [BOS_ID, ...promptIds];
  const out: number[] = [];
  let stoppedOnEos = false;

  for (let step = 0; step < maxNew; step++) {
    const window = context.slice(Math.max(0, context.length - ctx));
    const logits = model.lastLogits(window);
    for (const banned of GENERATION_BANNED_IDS) logits[banned] = -Infinity;
    assertUsableLogits(logits, V, step);

    let next: number;
    if (temperature === 0) {
      next = 0;
      let best = -Infinity;
      for (let i = 0; i < V; i++) {
        if (logits[i] > best) {
          best = logits[i];
          next = i;
        }
      }
    } else {
      let max = -Infinity;
      for (let i = 0; i < V; i++) if (logits[i] > max) max = logits[i];
      let sum = 0;
      const probs = new Float64Array(V);
      for (let i = 0; i < V; i++) {
        const e = Number.isFinite(logits[i]) ? Math.exp((logits[i] - max) / temperature) : 0;
        probs[i] = e;
        sum += e;
      }
      const u = rand() * sum;
      let acc = 0;
      next = V - 1;
      for (let i = 0; i < V; i++) {
        acc += probs[i];
        if (u < acc) {
          next = i;
          break;
        }
      }
    }

    // Structurally unreachable: every banned row is -inf and the logits are known finite,
    // so both branches must have chosen a real row. Asserted rather than filtered away —
    // this is the failure the `words` post-filter used to hide.
    if (BANNED.has(next)) {
      throw computeError(
        `generation selected the banned token ${SPECIAL_TOKENS[next] ?? next} at step ${step + 1}, ` +
          "which the -inf mask makes impossible; the model's arithmetic is not sound.",
      );
    }

    out.push(next);
    if (next === EOS_ID && stopAtEos) {
      stoppedOnEos = true;
      break;
    }
    context.push(next);
  }

  // One pass for both renderings, so they can never disagree. With stopAtEos=false,
  // `<eos>` is a line break rather than a terminator — the same rendering the backend
  // does, so both runtimes produce the same shape of verse. `words` is that same
  // sequence flattened; the `SPECIALS` test can only ever match `<eos>`, because every
  // other special is masked to -inf and its selection raises above.
  const words: string[] = [];
  const lines: string[][] = [[]];
  for (const w of vocab.decode(out)) {
    if (w === "<eos>") lines.push([]);
    else if (!SPECIALS.has(w)) {
      lines[lines.length - 1].push(w);
      words.push(w);
    }
  }
  const text = lines
    .filter((l) => l.length > 0)
    .map((l) => l.join(" "))
    .join("\n");
  return { ids: out, words, text, promptIds, promptOov, stoppedOnEos };
}

File: code/backend/src/llm_geometry/lex/generate.py
"""Generation from the Lexicon Lab's model — in-budget by construction.

The vocabulary **is** the budget, so there is no trie, no post-filter, and no way for an
out-of-budget word to appear: the only rows that could print something that is not a
budget word are the specials, and `GENERATION_BANNED_IDS` masks `<unk>`, `<bos>` and
`<pad>` to `-inf` before sampling. `<eos>` stays sampleable — it is how the model ends a
line. This is what makes the source's `mask_decode.py` defect (a trie that re-opened its
root mid-word and emitted `" hameat"`) structurally impossible here.

`temperature = 0` is greedy. The source clamped it to `1e-6` and always sampled, which
makes "show me this model's single most likely continuation" unavailable.

`<bos>` seeds an empty prompt, so that row is a real input the model has seen rather than
the source's dead embedding row (its `bos` variable actually held `<eos>`).
"""

from __future__ import annotations

from typing import Iterable, Sequence

import torch

from ..errors import ComputeError, InvalidParamError
from .config import (
    BOS_ID,
    DEFAULT_MAX_NEW_TOKENS,
    DEFAULT_SEED,
    DEFAULT_TEMPERATURE,
    EOS_ID,
    GENERATION_BANNED_IDS,
    MAX_NEW_TOKENS,
    SPECIAL_TOKENS,
)
from .model import LexModel
from .vocab import LexVocab, tokenize


def _banned_mask(vocab_rows: int, banned_ids: Iterable[int], device: torch.device) -> torch.Tensor:
    """Additive `-inf` mask over the banned rows, built once per call."""
    mask = torch.zeros(vocab_rows, device=device)
    for token_id in banned_ids:
        if not 0 <= int(token_id) < vocab_rows:
            raise InvalidParamError(f"banned id {token_id} is outside 0..{vocab_rows - 1}")
        mask[int(token_id)] = float("-inf")
    if not torch.isfinite(mask).any():
        raise InvalidParamError(
            "every vocabulary row is banned — there is nothing left to generate"
        )
    return mask


@torch.no_grad()
def generate_ids(
    model: LexModel,
    prompt_ids: Sequence[int] | None = None,
    *,
    max_new_tokens: int = DEFAULT_MAX_NEW_TOKENS,
    temperature: float = DEFAULT_TEMPERATURE,
    seed: int = DEFAULT_SEED,
    banned_ids: Sequence[int] = GENERATION_BANNED_IDS,
    stop_at_eos: bool = False,
) -> list[int]:
    """Continue `prompt_ids`; return the whole sequence (prompt included).

    `<eos>` ends a *line*, not the sample, so by default generation runs the full
    `max_new_tokens` and the caller renders line breaks. Pass `stop_at_eos=True` to get a
    single line — the model's own idea of where the line stops.

    Sampling is seeded, so a saved model reloaded from a bundle regenerates exactly the
    same text (SC-607). At `temperature == 0` the sampler is greedy and the seed is inert.
    """
    max_new_tokens = int(max_new_tokens)
    if not 1 <= max_new_tokens <= MAX_NEW_TOKENS:
        raise InvalidParamError(f"max_new_tokens must be in 1..{MAX_NEW_TOKENS}")
    temperature = float(temperature)
    if temperature < 0:
        raise InvalidParamError(f"temperature must be non-negative, got {temperature}")

    cfg = model.cfg
    ids = list(int(i) for i in (prompt_ids or []))
    if not ids:
        ids = [BOS_ID]  # a real input row, not a dead one
    for token_id in ids:
        if not 0 <= token_id < cfg.vocab_rows:
            raise InvalidParamError(f"prompt id {token_id} is outside 0..{cfg.vocab_rows - 1}")

    device = next(model.parameters()).device
    mask = _banned_mask(cfg.vocab_rows, banned_ids, device)
    gen = torch.Generator(device="cpu").manual_seed(int(seed))

    was_training = model.training
    model.eval()  # dropout must not perturb generation
    try:
        for _ in range(max_new_tokens):
            window = torch.tensor([ids[-cfg.ctx :]], dtype=torch.long, device=device)
            logits = model(window)[0, -1].float() + mask
            if not torch.isfinite(logits).any():
                raise ComputeError("all logits are -inf; generation cannot continue")
            if temperature == 0.0:
                next_id = int(torch.argmax(logits))
            else:
                probs = torch.softmax(logits / temperature, dim=-1)
                if not torch.isfinite(probs).all():
                    raise ComputeError(
                        "the model produced non-finite probabilities; refusing to sample"
                    )
                next_id = int(torch.multinomial(probs.cpu(), 1, generator=gen))
            ids.append(next_id)
            if stop_at_eos and next_id == EOS_ID:
                break
    finally:
        model.train(was_training)
    return ids


def generate_text(
    model: LexModel,
    vocab: LexVocab,
    prompt: str = "",
    *,
    max_new_tokens: int = DEFAULT_MAX_NEW_TOKENS,
    temperature: float = DEFAULT_TEMPERATURE,
    seed: int = DEFAULT_SEED,
    banned_ids: Sequence[int] = GENERATION_BANNED_IDS,
    stop_at_eos: bool = False,
) -> str:
    """Generate a continuation of `prompt` as words. Every word is in `vocab` by design.

    `<eos>` renders as a line break, which is what it means in a corpus of nursery rhymes.
    Out-of-budget prompt words become `<unk>` on the way IN — that is legitimate input, and
    the reported `<unk>` rate is exactly the measurable form of what a budget cannot say —
    but no special token can ever come back OUT.
    """
    if vocab.rows != model.cfg.vocab_rows:
        raise InvalidParamError(
            f"this vocabulary has {vocab.rows} rows but the model has "
            f"{model.cfg.vocab_rows}; they are not the same model"
        )
    prompt_ids = vocab.encode(tokenize(prompt))
    n_prompt = len(prompt_ids)
    ids = generate_ids(
        model,
        prompt_ids,
        max_new_tokens=max_new_tokens,
        temperature=temperature,
        seed=seed,
        banned_ids=banned_ids,
        stop_at_eos=stop_at_eos,
    )
    specials = set(SPECIAL_TOKENS)
    lines: list[list[str]] = [[]]
    for word in vocab.decode(ids[max(n_prompt, 1) :]):
        if word == "<eos>":
            lines.append([])
        elif word not in specials:
            lines[-1].append(word)
    return "\n".join(" ".join(line) for line in lines if line)
